src/order/rasiowa_sikorski.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2020 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import order.basic
+import data.equiv.encodable
+
+/-!
+# The Rasiowa–Sikorski lemma
+
+This file contains the (dual) Rasiowa–Sikorski lemma.
+
+## Main definitions
+
+We work with a preorder `P`, a term `p : P`, and a countable family `𝒟` of cofinal subsets of `P`.
+
+- `rasiowa_sikorski.witness p 𝒟 : set P`: the witness to the lemma, a `𝒟`-generic 'cofilter'.
+- `rasiowa_sikorski.directed_on p 𝒟`: the fact that the witness is (upwards) directed.
+- `rasiowa_sikorski.meets p 𝒟`: the fact that the witness is `𝒟`-generic.
+
+## Usage
+
+This provides an API for certain recursive constructions, similar to Zorn's lemma.
+Loosely speaking, suppose we want to construct an object satisfying some countable family
+of conditions. To apply Rasiowa–Sikorski, we need to:
+
+- Define a type `P`, whose terms should represent finitary attempts at constructing the object.
+- Provide a 'starting point' `p : P` for the construction.
+- Define a reflexive and transitive order `≤` on `P`, where `x ≤ y` should mean '`y` extends `x`'.
+- For each condition, define a subset of `P`, consisting of attemps which guarantee the condition,
+  together with a proof that any attempt can be extended so as to guarantee the condition.
+- Use `rasiowa_sikorski.witness` to define the desired object, as a limit of the attempts.
```

Comment:
Could you please wrap this into a theorem that illustrates this "usage"?

Reply:
This usage is precisely the lemma! I'd be happy to make the usage more convenient, but I'm not sure how. Would you prefer if the witness to the lemma and all its properties were wrapped together in one definition?

Reply:
Disclaimer: I never used this lemma and I don't know about "forcing with countable transitive models", so I may be wrong. When I was writing my comments I thought about an example that doesn't start with a type `P` but starts with an "object" and a countable family of conditions, then defines an auxiliary type `P` (using some sigma-type?) and applies the lemma.

If you think that it doesn't make sense, then feel free to just mark this conversation as resolved (see disclaimer above).

Reply:
I think I understand what you mean, but I can't see a general recipe for `P`. This is analogous to Zorn's lemma: if you want to construct an object with some property using Zorn's lemma, how do you know which preorder to consider?

I'll mark this as resolved, but would be interested to hear any suggestions on how to make the lemma easier to use.